kernel: Z_MEM_SLAB_INITIALIZER MACRO not compatible with C++

**Describe the bug**
It is not possible to define memory slabs using the MACRO `K_MEM_SLAB_DEFINE` in C++

**To Reproduce**
Just define a memory slab using the macro :
```cpp
K_MEM_SLAB_DEFINE(ipc_slab, IPC_TX_MESSAGE_SIZE, 8, 4);
```

**Expected behavior**
The compilation should succeed as for other kernel MACROs like `K_THREAD_DEFINE` or `K_MUTEX_DEFINE`.

**Impact**
The developper needs to define the structures by hand and initialize them at runtime (as described in the documentation https://docs.zephyrproject.org/1.9.0/kernel/memory/slabs.html#defining-a-memory-slab) : 

```cpp
struct k_mem_slab my_slab;
char __aligned(4) my_slab_buffer[6 * 400];
```

```cpp
void c_ipc::init(void) {
    k_mem_slab_init(&my_slab, my_slab_buffer, 400, 6);
}
```

**Screenshots or console output**

```
> Executing task: ninja <

[84/212] Building CXX object CMakeFiles/app.dir/src/app/c_ipc.cpp.obj
FAILED: CMakeFiles/app.dir/src/app/c_ipc.cpp.obj 
C:\gnuarmemb\bin\arm-none-eabi-g++.exe -DBUILD_VERSION=7dfd5458ce7a -DCORE_CM4 -DHSE_VALUE=25000000 -DKERNEL -DMBEDTLS_CONFIG_FILE=\"config-tls-generic.h\" -DSTM32F437xx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__LINUX_ERRNO_EXTENSIONS__ -D__PROGRAM_START -D__ZEPHYR__=1 -I../src -I../src/top -I../src/app -I../src/drv -I../src/fwk -I../src/cfg -I../zephyr/include -Izephyr/include/generated -I../zephyr/soc/arm/st_stm32/stm32f4 -I../zephyr/lib/libc/newlib/include -I../zephyr/drivers -I../zephyr/soc/arm/st_stm32/common -I../zephyr/subsys/net/lib/sockets/. -I../zephyr/subsys/net/lib/tls_credentials/. -I../zephyr/subsys/net/lib/dns/. -I../zephyr/subsys/net/ip -I../zephyr/subsys/net/lib/sockets -I../zephyr/subsys/net/l2 -I../modules/hal_stm32/stm32cube/stm32f4xx/soc -I../modules/hal_stm32/stm32cube/stm32f4xx/drivers/include -I../modules/hal_stm32/stm32cube/stm32f4xx/drivers/include/Legacy -I../modules/hal_stm32/stm32cube/common_ll/include -I../modules/cmsis/CMSIS/Core/Include -I../modules/segger/SEGGER -I../modules/segger/Config -I../zephyr/modules/segger/. -I../modules/mbedtls/mbedtls/include -I../zephyr/modules/mbedtls/configs -I../modules/fatfs/include -Os -fcheck-new -std=c++11 -Wno-register -fno-exceptions -imacros C:/Users/ldietrich/Documents/SauterECC2/app-ecc2_ctl-comm/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -fdiagnostics-color=always -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=hard -imacros C:/Users/ldietrich/Documents/SauterECC2/app-ecc2_ctl-comm/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -D__FIRMWARE_VERSION__=0x0008 -D__PRODUCT__=LET62WAPP -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=C:/Users/ldietrich/Documents/SauterECC2/app-ecc2_ctl-comm=CMAKE_SOURCE_DIR -fmacro-prefix-map=C:/Users/ldietrich/Documents/SauterECC2/app-ecc2_ctl-comm/zephyr=ZEPHYR_BASE -fmacro-prefix-map=C:/Users/ldietrich/Documents/SauterECC2/app-ecc2_ctl-comm=WEST_TOPDIR -ffunction-sections -fdata-sections -specs=nano.specs -MD -MT CMakeFiles/app.dir/src/app/c_ipc.cpp.obj -MF CMakeFiles\app.dir\src\app\c_ipc.cpp.obj.d -o CMakeFiles/app.dir/src/app/c_ipc.cpp.obj -c ../src/app/c_ipc.cpp   
In file included from ../zephyr/include/init.h:11,
                 from ../zephyr/include/device.h:29,
                 from ../zephyr/include/drivers/gpio.h:23,
                 from ../src/app/c_dfu.h:7,
                 from ../src/app/c_ipc.h:13,
                 from ../src/app/c_ipc.cpp:1:
../zephyr/include/kernel.h:4766:2: error: designator order for field 'k_mem_slab::wait_q' does not match declaration order in 'k_mem_slab'
 4766 |  }
      |  ^
../zephyr/include/kernel.h:4802:3: note: in expansion of macro 'Z_MEM_SLAB_INITIALIZER'
 4802 |   Z_MEM_SLAB_INITIALIZER(name, _k_mem_slab_buf_##name, \
      |   ^~~~~~~~~~~~~~~~~~~~~~
../src/app/c_ipc.cpp:51:1: note: in expansion of macro 'K_MEM_SLAB_DEFINE'
   51 | K_MEM_SLAB_DEFINE(ipc_slab, sizeof(c_ipc::tx_message_t), 8, 4);
      | ^~~~~~~~~~~~~~~~~
[93/212] Building CXX object CMakeFiles/app.dir/src/app/c_rest_server.cpp.obj
ninja: build stopped: subcommand failed.
The terminal process "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -Command ninja" terminated with exit code: 1.

Press any key to close the terminal.
```

**Environment**
- Windows 10 Professionnel 20H2
- Zephyr  version : 2.6.0
- Ninja 1.10.2, 
- arm-none-eabi-g++.exe (GNU Arm Embedded Toolchain 10-2020-q4-major) 10.2.1 20201103 (release)

**Additional context**

The problem is very similar to what is described in this issue : https://github.com/zephyrproject-rtos/zephyr/issues/14540

The MACRO used to define a the memory slab structure `struct k_mem_slab` declares members in a different order, see :

https://github.com/zephyrproject-rtos/zephyr/blob/aba8a21001ffd2a0084085fc29cf6db7e3c51ce1/include/kernel.h#L4758-L4782

I'm able to make the compilation work by reordering the initializers in the MACRO `K_MEM_SLAB_DEFINE` like this :

```cpp
#define Z_MEM_SLAB_INITIALIZER(obj, slab_buffer, slab_block_size, \
			       slab_num_blocks) \
	{ \
	.wait_q = Z_WAIT_Q_INIT(&obj.wait_q), \
	.lock = {}, \
	.num_blocks = slab_num_blocks, \
	.block_size = slab_block_size, \
	.buffer = slab_buffer, \
	.free_list = NULL, \
	.num_used = 0, \
	}
```

--- a/include/kernel.h
+++ b/include/kernel.h
@@ -4772,8 +4772,8 @@ struct k_mem_slab {
 #define Z_MEM_SLAB_INITIALIZER(obj, slab_buffer, slab_block_size, \
 			       slab_num_blocks) \
 	{ \
-	.lock = {}, \
 	.wait_q = Z_WAIT_Q_INIT(&obj.wait_q), \
+	.lock = {}, \
 	.num_blocks = slab_num_blocks, \
 	.block_size = slab_block_size, \
 	.buffer = slab_buffer, \

